RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc

```diff
@@ -10,44 +11,430 @@
+  for (auto &itvl : m_coverage) {
+    itvl.x = std::sqrt(itvl.x);
+    itvl.y = std::sqrt(itvl.y);
+  }
+}
+
+bool MkFitGeometryESProducer::GapCollector::find_gap(Interval &itvl, float eps) {
+  auto i = m_coverage.begin();
+  while (i != m_coverage.end()) {
+    auto j = i;
+    ++j;
+    if (j != m_coverage.end()) {
+      if (j->x - i->y > eps) {
+        itvl = {i->y, j->x};
+        return true;
+      }
+      i = j;
+    } else {
+      break;
+    }
+  }
+  return false;
+}
+
+void MkFitGeometryESProducer::GapCollector::print_gaps() {
+  auto i = m_coverage.begin();
+  while (i != m_coverage.end()) {
+    auto j = i;
+    ++j;
+    if (j != m_coverage.end()) {
+      printf("(%f, %f)->%f ", i->y, j->x, j->x - i->y);
+      i = j;
+    } else {
+      break;
+    }
+  }
+  printf("\n");
+}
+
+//------------------------------------------------------------------------------
+
+void MkFitGeometryESProducer::considerPoint(const GlobalPoint &gp, mkfit::LayerInfo &li) {
+  // Use radius squared during bounding-region search.
+  float r = gp.perp2(), z = gp.z();
+  li.extend_limits(r, z);
+}
+
+void MkFitGeometryESProducer::fillShapeAndPlacement(const GeomDet *det,
+                                                    mkfit::TrackerInfo &trk_info,
+                                                    layer_gap_map_t *lgc_map) {
+  DetId detid = det->geographicalId();
+
+  float xy[4][2];
+  float dz;
+  const Bounds *b = &((det->surface()).bounds());
+
+  if (const TrapezoidalPlaneBounds *b2 = dynamic_cast<const TrapezoidalPlaneBounds *>(b)) {
+    // Trapezoidal
+    std::array<const float, 4> const &par = b2->parameters();
+
+    // These parameters are half-lengths, as in CMSIM/GEANT3
+    // https://github.com/trackreco/cmssw/blob/master/Fireworks/Core/src/FWGeometry.cc#L241
+    // https://github.com/root-project/root/blob/master/geom/geom/src/TGeoArb8.cxx#L1331
+    /*
+    i.shape[0] = 1;
+    i.shape[1] = par[0];  // hBottomEdge - dx1
+    i.shape[2] = par[1];  // hTopEdge    - dx2
+    i.shape[3] = par[2];  // thickness   - dz
+    i.shape[4] = par[3];  // apothem     - dy1
+
+          geoShape = new TGeoTrap(info.shape[3],  //dz
+                            0,              //theta
+                            0,              //phi
+                            info.shape[4],  //dy1
+                            info.shape[1],  //dx1
+                            info.shape[2],  //dx2
+                            0,              //alpha1
+                            info.shape[4],  //dy2
+                            info.shape[1],  //dx3
+                            info.shape[2],  //dx4
+                            0);             //alpha2
+
+      TGeoTrap::TGeoTrap(Double_t dz, Double_t theta, Double_t phi, Double_t h1,
+               Double_t bl1, Double_t tl1, Double_t alpha1, Double_t h2, Double_t bl2,
+               Double_t tl2, Double_t alpha2)
+          :TGeoArb8("", 0, 0)
+    {
```

Comment:
is this intended as documentation, or is it leftover?

Reply:
It's documentation ... it was quite fun to figure out how to interpret the Bounds objkect parameter array. I can move it to a file in doc/ and reference it.

Reply:
moved to doc/reco-geom-notes.txt